internal/daemon/api_files.go

```diff
@@ -0,0 +1,627 @@
+		}
+	}
+	return syncResponseWithError(result, firstErr)
+}
+
+func writeFile(item writeFilesItem, source io.Reader) error {
+	if !path.IsAbs(item.Path) {
+		return nonAbsolutePathError(item.Path)
+	}
+
+	// Create parent directory if needed
+	if item.MakeDirs {
+		err := os.MkdirAll(path.Dir(item.Path), 0o755)
+		if err != nil {
+			return fmt.Errorf("creating directory: %w", err)
+		}
+	}
+
+	// Create file and write contents to temporary file.
+	perm, err := parsePermissions(item.Permissions, 0o644)
+	if err != nil {
+		return err
+	}
+	tempPath := item.Path + ".pebble-temp"
+	f, err := os.OpenFile(tempPath, os.O_RDWR|os.O_CREATE|os.O_TRUNC, perm)
+	if err != nil {
+		return fmt.Errorf("creating file: %w", err)
+	}
+	_, err = io.Copy(f, source)
+	if err != nil {
+		_ = f.Close()
+		_ = os.Remove(tempPath)
+		return fmt.Errorf("writing file: %w", err)
+	}
+	err = f.Close()
+	if err != nil {
+		_ = os.Remove(tempPath)
+		return fmt.Errorf("closing file: %w", err)
+	}
+
+	// Update user and group if necessary.
+	err = updateUserAndGroup(tempPath, item.UserID, item.User, item.GroupID, item.Group)
+	if err != nil {
+		_ = os.Remove(tempPath)
+		return fmt.Errorf("setting user and group: %w", err)
+	}
+
+	// Atomically move temporary file to final location.
+	err = os.Rename(tempPath, item.Path)
+	if err != nil {
+		return fmt.Errorf("moving temporary file to path: %w", err)
+	}
+	return nil
+}
+
+func parsePermissions(permissions string, defaultMode os.FileMode) (os.FileMode, error) {
+	if permissions == "" {
+		return defaultMode, nil
+	}
+	perm, err := strconv.ParseUint(permissions, 8, 32)
+	if err != nil || len(permissions) != 3 {
+		return 0, fmt.Errorf("permissions must be a 3-digit octal string, not %q", permissions)
+	}
+	return os.FileMode(perm), nil
+}
+
+// Creating directories
+
+type makeDirsItem struct {
+	Path        string `json:"path"`
+	MakeParents bool   `json:"make-parents"`
+	Permissions string `json:"permissions"`
+	UserID      int    `json:"user-id"`
+	User        string `json:"user"`
+	GroupID     int    `json:"group-id"`
+	Group       string `json:"group"`
+}
+
+func makeDirs(dirs []makeDirsItem) Response {
+	result := make([]fileResult, len(dirs))
+	var firstErr error
+	for i, dir := range dirs {
+		err := makeDir(dir)
+		if err != nil && firstErr == nil {
+			firstErr = err
+		}
+		result[i] = fileResult{
+			Path:  dir.Path,
+			Error: fileErrorToResult(err),
+		}
+	}
+	return syncResponseWithError(result, firstErr)
+}
+
+func makeDir(dir makeDirsItem) error {
+	if !path.IsAbs(dir.Path) {
+		return nonAbsolutePathError(dir.Path)
+	}
+	perm, err := parsePermissions(dir.Permissions, 0o755)
+	if err != nil {
+		return err
+	}
+	if dir.MakeParents {
+		err = os.MkdirAll(dir.Path, perm)
+	} else {
+		err = os.Mkdir(dir.Path, perm)
+	}
+	if err != nil {
+		return err
+	}
+	err = updateUserAndGroup(dir.Path, dir.UserID, dir.User, dir.GroupID, dir.Group)
+	if err != nil {
+		return fmt.Errorf("setting user and group: %w", err)
```

Comment:
Same as earlier note. Not sure if I missed one of these so would be worth double checking.

Reply:
Done.